src/utils/configuration.ts

```diff
@@ -0,0 +1,164 @@
+		try {
+			const clients = await getClients();
+			const client = clients?.find((client) => client.id === CLIENT);
+
+			if (client && clients.length > 1) {
+				await redisClient.lRem(REDIS_CONFIG_KEY, 1, JSON.stringify(client));
+				if (client.isWriting) {
+					await promote();
+				}
+			} else if (client && clients.length === 1) {
+				await redisClient.delete(REDIS_CONFIG_KEY);
+				throw new Error(`No event publishing clients are running`);
+			}
+		} catch (error) {
+			logger.alert(`Failed to shutdown gracefully: ${error.message}`);
+		}
+	};
+
+	const awaitPromotion = async ({ callback }) => {
+		await redisClient.subscribe(REDIS_CLIENT_CHANNEL);
+
+		redisClient.forceGetClient().on('message', async (channel, message) => {
+			if (channel === REDIS_CLIENT_CHANNEL) {
+				const { isWriting = false, isFailover = false } = JSON.parse(message);
+				if (isWriting) {
+					logger.info(`Starting to write on client: ${CLIENT}`);
+					callback();
+				}
+				if (isFailover) throw new Error(`Failing over client: ${CLIENT}`);
+			}
+		});
+
+		redisClient.forceGetClient().on('error', (e) => {
+			throw new Error(`Redis Error: ${e.message}`);
+		});
+
+		redisClient.forceGetClient().on('end', () => {
+			throw new Error('Redis connection was closed');
+		});
+	};
+
+	const checkIfShouldPromote = async () => {
+		const clients = await getClients();
+		const clientIsWriting = clients?.find((client) => client.isWriting);
+		if (!clientIsWriting) {
+			await promote();
+		}
+	};
+
+	const checkIfClientsConnected = async () => {
+		const connectedClients = [CLIENT];
+
+		await redisClient.subscribe(REDIS_CHANNEL_PREFIX);
+
+		// On startup a new client will ping existing clients to determine if the
+		// connections are still connected
+		redisClient.forceGetClient().on('message', async (channel, message) => {
+			if (channel === REDIS_CHANNEL_PREFIX) {
+				const { id, ping = false, pong = false } = JSON.parse(message);
+
+				if (id === CLIENT && ping) {
+					await redisClient.publish(`${REDIS_CHANNEL_PREFIX}`, {
+						id,
+						pong: true,
+					});
+				}
+
+				if (id !== CLIENT && pong) {
+					connectedClients.push(id);
+				}
+			}
+		});
+
+		const clients = await getClients();
+
+		clients
+			?.filter((client) => client.id !== CLIENT)
+			.map(async (client) => {
+				await redisClient.publish(`${REDIS_CHANNEL_PREFIX}`, {
+					id: client.id,
+					ping: true,
+				});
+			});
+
+		// Clean up any stale connections
+		setTimeout(async () => {
+			logger.info(`Connected clients: ${connectedClients}`);
+			const disconnectedClients = clients.filter(
+				(client) => !connectedClients.includes(client.id)
+			);
+			logger.info(`Disconnected clients: ${disconnectedClients}`);
+			disconnectedClients.map(async (client) => {
+				await redisClient.lRem(REDIS_CONFIG_KEY, 1, JSON.stringify(client));
+			});
+
+			await checkIfShouldPromote();
+		}, 3000);
+	};
+
+	const promote = async () => {
```

Comment:
should make sure that this is idempotent -- if there's a race condition and multiple clients are trying to promote and send a deuplicate message, is that okay?